### PR TITLE
docs: Add detailed contributing guidelines

### PR DESCRIPTION
**Pull request by rata**
This commits adds some guidelines that I believe will help to review PRs
and simplify the PR creation process, that is currently a little more
chaotic than it could be.

Hopefully PRs will be merged faster if they are easier to review too.

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+
+ * Your development history is not important. Use git rebase to squash and order
+   commits in a way that is easy to review. Just keep the useful commits and not
+   your development history that led you to the change.
+
+ * Consider reviewing yourself the changes before opening a PR. Most of the
+   times, you can catch several errors.
+
+ * Use the PR description as a cover letter and give the context you think
+   people reviewing the code might need. Use the PR description to give an
```
Comment (johananl) [inline docs/CONTRIBUTING.md:82]:
```suggestion
   reviewers might need. Use the PR's description to give an
```

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   overview, ask the questions about yet unsolved things in your PR, list
```
Comment (johananl) [inline docs/CONTRIBUTING.md:83]:
```suggestion
   overview, raise questions about yet-unresolved issues in your PR, list
```

File: docs/CONTRIBUTING.md [hunk 3]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   pending/TODO items, etc.
```
Comment (johananl) [inline docs/CONTRIBUTING.md:84]:
```suggestion
   pending/TODO items etc.
```

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   added. Consider to just squash the changes.
+
+ * Your development history is not important. Use git rebase to squash and order
+   commits in a way that is easy to review. Just keep the useful commits and not
+   your development history that led you to the change.
+
+ * Consider reviewing yourself the changes before opening a PR. Most of the
+   times, you can catch several errors.
+
+ * Use the PR description as a cover letter and give the context you think
```
Comment (johananl) [inline docs/CONTRIBUTING.md:81]: IMO the most important thing to include in the description is the rationale for opening the PR: **why** I'm proposing the change and think it's a good idea to merge it.

File: docs/CONTRIBUTING.md [hunk 5]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   people reviewing the code might need. Use the PR description to give an
+   overview, ask the questions about yet unsolved things in your PR, list
+   pending/TODO items, etc.
+
+If PRs follow these rules, usually, they are more easy to review and merge.
```
Comment (johananl) [inline docs/CONTRIBUTING.md:86]:
```suggestion
PRs which follow these rules are easier to review and merge.
```

File: docs/CONTRIBUTING.md [hunk 6]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+
+It is totally okay to bend the rules sometimes. For example, if you reworked a
+piece of code and you realize it happens to fix other issues, that is totally
+fine. A strict interpretation of the "commits shouls be atomic" rule might make
+you think you need to fix the issues in the original code and _then_ do your
+rework that possibly deletes the fixes you created. That is not the idea, we
+don't want to create pointless overhead. Just do your rework and mention that it
+also fixes some other bugs that were present before your PR. Don't add code you
+will later delete (as another rule says).
```
Comment (johananl) [inline docs/CONTRIBUTING.md:95]: I would remove this paragraph. I think all it does is weaken the message we are trying to convey with these guidelines. People know guidelines aren't laws and we should trust them to use their good judgement when contributing.

File: docs/CONTRIBUTING.md [hunk 7]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+
+Use your own judgement to decide when is best to follow these rules. Don't make
+your life too hard, keep the commit history useful and try to make the review
+process easy, this is the intent after all.
```
Comment (johananl) [inline docs/CONTRIBUTING.md:99]: I would remove this paragraph, too. It doesn't make the message clearer IMO.

File: docs/CONTRIBUTING.md [hunk 8]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   explained in the the Linux Kernel [submitting patches][linux-sep-changes]
+   guide.
+
+ * Commit messages should explain the intention, _why_ something is done. This,
+   too, is best explained in [this section][linux-desc-changes] from the Linux
+   Kernel submitting patches guide.
+
+ * Commit titles should be meaningful and describe _what_ the commit does
+   ("refactor" and "fix" would be the title of ~all commits otherwise). For
+   example, when you are using "refactor" in a title, you probably include a list
```
[the remark this answers is not in the document]
Comment (rata) [reply docs/CONTRIBUTING.md:66]: My intention was the oposite: if you find using it, consider spliting the changes :)

File: docs/CONTRIBUTING.md [hunk 9]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   of things in the commit message. Consider doing those things in different
+   commits instead. Commits should be atomic.
+
+ * Don't add code you will change later (it makes pointless to review that
```
[the remark this answers is not in the document]
Comment (rata) [reply docs/CONTRIBUTING.md:70]: A later commit, will fix. Thanks!

File: docs/CONTRIBUTING.md (hunk 3 above)
Comment (rata) [reply docs/CONTRIBUTING.md:84]: Why not `,`?

File: docs/CONTRIBUTING.md (hunk 7 above)
Comment (rata) [reply docs/CONTRIBUTING.md:99]: Okay

File: docs/CONTRIBUTING.md (hunk 6 above)
Comment (rata) [reply docs/CONTRIBUTING.md:95]: I wanted to emphasize that you shouldn't have to do a lot of extra work "to make your PR according to the rules". I thought that interpretation I used in the example was _plausible_ and wanted to make clear it was not the intention.

But if it clear for others too, maybe I'll remove it. Thanks! :)

**Comment by rata** [comment]
> Thanks for the PR +1 Added some comments.

Thanks, will fix the rest next week. And thanks a lot for the commit suggestions, it makes it way more easy (to avoid misunderstandings and to incorporate a change :))

File: docs/CONTRIBUTING.md [hunk 10]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
-might have made in this commit. Changes to unrelated parts of the
-codebase should be kept as separate commits.
+## Authoring PRs
+
+These are general guidelines for making PRs/commits easier to review. Use your own
+judgement to decide when it is best to follow them. These are not strict rules
+but highly recommended to follow:
+
+ * Commits should be atomic and self-contained. Divide logically separate changes
+   to separate commits. This principle  is best explained in the the Linux Kernel
```
Comment (invidian) [inline docs/CONTRIBUTING.md:56]: Any specific reason for double space here?
```suggestion
   to separate commits. This principle is best explained in the the Linux Kernel
```

File: docs/CONTRIBUTING.md (hunk 3 above)
Comment (invidian) [reply docs/CONTRIBUTING.md:84]: https://english.stackexchange.com/a/1470 seems like something which should go into style guide.

File: docs/CONTRIBUTING.md (hunk 10 above)
Comment (rata) [reply docs/CONTRIBUTING.md:56]: Nope :see_no_evil:

File: docs/CONTRIBUTING.md (hunk 3 above)
Comment (rata) [reply docs/CONTRIBUTING.md:84]: I'm not sure I follow why that is answer. You are trating ' /` as or? I can remove pending, keep it `list TODO items` and that will make everyone happy?

File: docs/CONTRIBUTING.md [hunk 11]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
-## Commit guidelines
-
-The title of the commit message should describe the _what_ about the
-changes. Additionally, it is helpful to add the _why_ in the body of
-the commit changes. Make sure to include any assumptions that you
-might have made in this commit. Changes to unrelated parts of the
-codebase should be kept as separate commits.
+## Contributing PRs
+
+This is a general guideline to make PRs/commits easier to review. Use your own
```
[the remark this answers is not in the document]
Comment (rata) [reply docs/CONTRIBUTING.md]: Okay, removed. Thanks!

File: docs/CONTRIBUTING.md (hunk 6 above)
Comment (rata) [reply docs/CONTRIBUTING.md:95]: Okay, removed as @invidian said he agreed with the comments you did too :)

File: docs/CONTRIBUTING.md (hunk 4 above)
Comment (rata) [reply docs/CONTRIBUTING.md]: Added, thanks!

File: docs/CONTRIBUTING.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (rata) [reply docs/CONTRIBUTING.md]: Fixed

File: docs/CONTRIBUTING.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (rata) [reply docs/CONTRIBUTING.md:65]: Just removed the example. Hope it is clearer now, with less text :)

File: docs/CONTRIBUTING.md [hunk 12]
```diff
@@ -46,13 +46,60 @@ make update-assets
...
+   tests, _logically_ different changes in different commits). This is best
+   explained in the the Linux Kernel [submitting patches][linux-sep-changes]
+   guide.
+
+ * Commit messages should explain the intention, _why_ something is done. This,
+   too, is best explained in [this section][linux-desc-changes] from the Linux
+   Kernel submitting patches guide.
+
+ * Commit titles should be meaningful and describe _what_ the commit does
+   ("refactor" and "fix" would be the title of ~all commits otherwise). For
```
[the remark this answers is not in the document]
Comment (rata) [reply docs/CONTRIBUTING.md:64]: Removed, thanks!

**Comment by rata** [comment]
@invidian @johananl PTAL :)

File: docs/CONTRIBUTING.md [hunk 13]
```diff
@@ -46,13 +46,43 @@ make update-assets
...
+ * Commit messages should explain the intention, _why_ something is done. This,
+   too, is best explained in [this section][linux-desc-changes] from the Linux
+   Kernel patch submission guide.
+
+ * Commit titles (the first line in a commit) should be meaningful and describe
+   _what_ the commit does.
+
+ * Don't add code you will change in a later commit (it makes it pointless to
+   review that commit), nor create a commit to add code an earlier commit should
+   have added. Consider to just squash the changes.
```
Comment (johananl) [inline docs/CONTRIBUTING.md:66]:
```suggestion
   have added. Consider squashing the relevant commits instead.
```

File: docs/CONTRIBUTING.md [hunk 14]
```diff
@@ -46,13 +46,43 @@ make update-assets
...
+
+ * It's not important to retain your development history when contributing a
+   change. Use `git rebase` to squash and order commits in a way that is easy to
```
Comment (johananl) [inline docs/CONTRIBUTING.md:69]:
```suggestion
   change. Use `git rebase` to squash and order commits in a way that makes them easy to
```

File: docs/CONTRIBUTING.md [hunk 15]
```diff
@@ -46,13 +46,43 @@ make update-assets
...
+## Authoring PRs
 
-The title of the commit message should describe the _what_ about the
-changes. Additionally, it is helpful to add the _why_ in the body of
-the commit changes. Make sure to include any assumptions that you
-might have made in this commit. Changes to unrelated parts of the
-codebase should be kept as separate commits.
+These are general guidelines for making PRs/commits easier to review:
+
+ * Commits should be atomic and self-contained. Divide logically separate changes
```
Comment (johananl) [inline docs/CONTRIBUTING.md:53]:
```suggestion
 * Commits should be atomic and self contained. Divide logically-separate changes
```

AFAIK hyphens in compound adjectives should be used when the adjective comes *after* the verb. Example:

:heavy_check_mark: This example is **self contained**.
:heavy_check_mark: I'm writing a **self-contained** example.

File: docs/CONTRIBUTING.md [hunk 16]
```diff
@@ -46,13 +46,43 @@ make update-assets
...
+   that led to the final state.
+
+ * Consider reviewing the changes yourself before opening a PR. It is likely
+   you will catch errors when looking at your code critically and thus save the
+   reviewers (and yourself) time.
+
+ * Use the PR's description as a "cover letter" and give the context you think
+   reviewers might need. Use the PR's description to explain why you are
+   proposing the change, give an overview, raise questions about yet-unresolved
+   issues in your PR, list TODO items, etc.
```
Comment (johananl) [inline docs/CONTRIBUTING.md:80]:
```suggestion
   issues in your PR, list TODO items etc.
```

That's an Oxford comma ;-)

File: docs/CONTRIBUTING.md (hunk 15 above)
Comment (rata) [reply docs/CONTRIBUTING.md:53]: Good to know! I always did what my gut feeling told me :see_no_evil:

File: docs/CONTRIBUTING.md (hunk 16 above)
Comment (rata) [reply docs/CONTRIBUTING.md:80]: Oh, right. But it feels weird without an `and` or `or`. Is it really used like this in this (just space and `etc.`)? My eyes hurt a little, but if it's correct...

Comment (rata) [reply docs/CONTRIBUTING.md:80]: Added it, I don't like it but if there is a less ugly way to do it, let's open another PR and get done with this one. This is not so important and is the only thing pending :)

**Comment by rata** [comment]
ugh, github doesn't allow me to merge with just squashing your last small suggestions. Can one re-approve this? @invidian @johananl 

It is the same that was already LGTM, just squashed the suggestions into the original commit and pushed.

**Comment by rata** [comment]
Thanks again! :)

**Pull request merged**